### PR TITLE
Engine flow control support

### PR DESCRIPTION
**Pull request by vijaynalawade**
**What kind of change does this PR introduce?** (check one with "x")
```
[] Bugfix
[x] Feature
[] Code style update (formatting, local variables)
[] Refactoring (no functional changes, no api changes)
[] Other... Please describe:
```

**Fixes**: #
https://github.com/project-flogo/core/issues/226
**What is the current behavior?**
Today, engine will simply accept as many events as possible(Direct Runner) or block trigger go-routines when worker queue is full(Pooled Runner). This could lead to a performance issue or even loss of events due to resource constrains like CPU and Memory or blocked workers(Pooled runner). This might also lead to engine crash. 
**What is the new behavior?**
Added new interface that Trigger implementations can implement to support pause/resume.  Also added a mechanism that can force pausing/resuming of triggers. The flow control policies are in not scope of this PR.

### CODE REVIEW

**Comment by lixingwang** [comment]
@vijaynalawade I am thining the controller should be app-specific, can we move it as an method of app. such as: `app.SetController(defaultController)` etc. 

Users may don't need a controller for most cases.  so can we add an env var to control it? by default, there is no controller
@mellistibco  thoughts?

File: app/controller.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+func setController(app *App) {
+	controller = controllerData{}
+	controller.notify = make(chan bool)
+	controller.triggers = make(map[string]trigger.FlowControlAware)
+	for id, trgW := range app.triggers {
+		if t, ok := trgW.trg.(trigger.FlowControlAware); ok {
+			controller.triggers[id] = t
+		}
+	}
+	go controller.startController()
```
Comment (mellistibco) [inline app/controller.go:44]: if you have no flow control aware triggers then why start the controller?

File: app/controller.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// Resume triggers
+func (c controllerData) resumeTriggers() error {
+	// Resume  triggers
+	log.RootLogger().Info("Resuming Triggers...")
+	for _, trg := range c.triggers {
+		err := trg.Resume()
```
Comment (mellistibco) [inline app/controller.go:52]: debug log to say which triggers have been resumed?

**Review by mellistibco — CHANGES_REQUESTED** [review]

**Comment by mellistibco** [comment]
agree with @lixingwang. I think in most scenarios the user will not use this...

File: app/controller.go (hunk 1 above)
Comment (vijaynalawade) [reply app/controller.go]: @mellistibco Fixed in latest update.

File: app/controller.go (hunk 2 above)
Comment (vijaynalawade) [reply app/controller.go:54]: @mellistibco Fixed in latest update.

**Comment by vijaynalawade** [comment]
@mellistibco @lixingwang Please review latest update.

**Review by lixingwang — APPROVED** [review]
Looks good to me

**Review by mellistibco — APPROVED** [review]

**Pull request merged**